RFS/src/main/java/com/rfs/common/OpenSearchClient.java

```diff
@@ -257,20 +280,54 @@ public Optional<ObjectNode> getSnapshotStatus(
         }
     }
 
-    public Mono<BulkResponse> sendBulkRequest(String indexName, String body, IRfsContexts.IRequestContext context) {
-        String targetPath = indexName + "/_bulk";
+    Retry getBulkRetryStrategy() {
+        return bulkRetryStrategy;
+    }
 
-        return client.postAsync(targetPath, body, context)
-            .map(response -> new BulkResponse(response.statusCode, response.statusText, response.headers, response.body))
-            .flatMap(resp -> {
-                if (resp.hasBadStatusCode() || resp.hasFailedOperations()) {
-                    logger.error(resp.getFailureMessage());
+    public Mono<BulkResponse> sendBulkRequest(String indexName, List<BulkDocSection> docs, IRfsContexts.IRequestContext context) {
+        final var docsMap = docs.stream().collect(Collectors.toMap(d -> d.getDocId(), d -> d));
+        return Mono.defer(() -> {
+            final String targetPath = indexName + "/_bulk";
+            log.atTrace()
+                .setMessage("Creating bulk body with document ids {}")
+                .addArgument(() -> docsMap.keySet())
+                .log();
+            var body = BulkDocSection.convertToBulkRequestBody(docsMap.values());
+            return client.postAsync(targetPath, body, context)
+                .flatMap(response -> {
+                    var resp = new BulkResponse(response.statusCode, response.statusText, response.headers, response.body);
+                    if (!resp.hasBadStatusCode() && !resp.hasFailedOperations()) {
+                        return Mono.just(resp);
+                    }
+                    // Remove all successful documents for the next bulk request attempt
+                    var successfulDocs = resp.getSuccessfulDocs();
+                    successfulDocs.forEach(docsMap::remove);
+                    log.atWarn()
+                        .setMessage("After bulk request on index '{}', {} more documents have succeed, {} remain")
+                        .addArgument(indexName)
+                        .addArgument(successfulDocs::size)
+                        .addArgument(docsMap::size)
+                        .log();
                     return Mono.error(new OperationFailed(resp.getFailureMessage(), resp));
```

Comment:
I'm curious why we're removing successful docs rather than making new request out of the ones that failed.  Is this a bet that, on the whole, there will be more failed docs than successful ones when there are any failed docs?  Also, and I could definitely be wrong here, it feels more efficient to construct a new array from a set of elements of known size rather than iteratively remove elements from an existing array.  Both aspects lead me to intuit that we might be better off taking the alternative approach.

What are your thoughts?

Reply:
In the happy path all docs succeed and we don't get to the winnowing code.

For the 'some docs failed case' Its easier to determine that a doc has been successfully uploaded - whereas if we do not get a full list of documents in the bulk response we need work backwards from the list of doc as we scan and make sure it isn't absent.  I'm optimizing this code towards reliability.

Reply:
I'd love to take a look at some examples with you because it sounds like we understand the response format differently.  From my reading of the spec [1], we end up with an entry for every element of the request and it appears easy to check if any given element failed.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/7.10/docs-bulk.html

Reply:
Per conversation - we don't have evidence to indicate how many failed requests will show up in a response when some do (e.g. is it usually a lot or very few?).  Additionally, the `docsMap` doesn't contain the bulk sections itself, it just contains references to them.  As a result, while I feel intuitively that we'd be better off making a new map using the deleted docs, the impact appears uncertain enough to keep the current approach and see how it works out.